File: hw/cevallos/u3/hw29Singleton/python/controllers/sales_controller.py
from typing import List, Dict, Any
from models.sold_thing import SoldThing
from models.discount_singleton import DiscountSingleton


class SalesController:
    """Controller for managing sales operations."""
    
    def __init__(self) -> None:
        self._sales: List[SoldThing] = []
        self._discount_singleton = DiscountSingleton()
    
    def create_sale(self, item_name: str, price: float) -> SoldThing:
        """Create a new sale record."""
        if price <= 0:
            raise ValueError("Price must be greater than zero.")
        
        sold_item = SoldThing(name=item_name, original_price=float(price))
        self._sales.append(sold_item)
        return sold_item
    
    def get_all_sales(self) -> List[SoldThing]:
        """Get all sales records."""
        return self._sales.copy()
    
    def get_sales_summary(self) -> List[Dict[str, Any]]:
        """Get formatted sales data for display."""
        summary = []
        
        for sale in self._sales:
            summary.append({
                'id': sale.sale_id,
                'name': sale.name,
                'original_price': sale.original_price,
                'discount_percentage': sale.get_discount_percentage(),
                'discount_amount': sale.get_discount_amount(),
                'final_price': sale.calculate_final_price(),
                'sale_date': sale.sale_date.strftime('%Y-%m-%d %H:%M:%S')
            })
        
        return summary
    
    def update_discount_percentage(self, new_discount: float) -> bool:
        """Update the global discount percentage."""
        try:
            return self._discount_singleton.set_discount_percentage(new_discount)
        except ValueError as e:
            raise e
    
    def get_current_discount(self) -> float:
        """Get current discount percentage."""
        return self._discount_singleton.get_discount_percentage()
    
    def clear_sales(self) -> None:
        """Clear all sales records."""
        self._sales.clear()

File: hw/cevallos/u3/hw29Singleton/JavaScript/src/controllers/SalesController.js
const SoldThing = require('../models/soldThing');
const DiscountSingleton = require('../models/DiscountSingleton');

class SalesController {
  constructor() {
    this.sales = [];
    this.discountSingleton = DiscountSingleton.getInstance();
  }

  createSale(itemName, price) {
    try {
      const soldItem = new SoldThing(itemName, price);
      this.sales.push(soldItem);
      return soldItem;
    } catch (error) {
      throw new Error(`Failed to create sale: ${error.message}`);
    }
  }

  getAllSales() {
    return [...this.sales];
  }

  getSalesSummary() {
    return this.sales.map(sale => sale.toJSON());
  }

  updateDiscountPercentage(newDiscount) {
    try {
      return this.discountSingleton.setDiscountPercentage(newDiscount);
    } catch (error) {
      throw new Error(`Failed to update discount: ${error.message}`);
    }
  }

  getCurrentDiscount() {
    return this.discountSingleton.getDiscountPercentage();
  }

  clearSales() {
    this.sales = [];
    return this.sales.length === 0;
  }

  getSalesStatistics() {
    const summary = this.getSalesSummary();
    
    if (summary.length === 0) {
      return {
        totalSales: 0,
        totalRevenue: 0,
        totalDiscount: 0,
        averageDiscount: 0
      };
    }

    const totalRevenue = summary.reduce((sum, sale) => sum + sale.finalPrice, 0);
    const totalDiscount = summary.reduce((sum, sale) => sum + sale.discountAmount, 0);
    const averageDiscount = totalDiscount / summary.length;

    return {
      totalSales: summary.length,
      totalRevenue: parseFloat(totalRevenue.toFixed(2)),
      totalDiscount: parseFloat(totalDiscount.toFixed(2)),
      averageDiscount: parseFloat(averageDiscount.toFixed(2))
    };
  }
}

module.exports = SalesController;